File: libs/ZmqLib/go/sub/subscriber.go
package subscriber

import (
	"io/ioutil"

	"github.com/golang/glog"
	zmq "github.com/pebbe/zmq4"
)

// loadCertificates is used to load all required certs
func loadCertificates(publicKeysDir string, privateKeysDir string) (string, string, string, string) {
	clientPublicKey, err := ioutil.ReadFile(publicKeysDir + "/client.key")
	if err != nil {
		glog.Errorf("Error reading certs : %v", err)
		return "", "", "", ""
	}
	clientPublic := string(clientPublicKey)

	clientSecretKey, err := ioutil.ReadFile(privateKeysDir + "/client.key_secret")
	if err != nil {
		glog.Errorf("Error reading certs : %v", err)
		return "", "", "", ""
	}
	clientSecret := string(clientSecretKey)

	serverPublicKey, err := ioutil.ReadFile(publicKeysDir + "/server.key")
	if err != nil {
		glog.Errorf("Error reading certs : %v", err)
		return "", "", "", ""
	}
	serverPublic := string(serverPublicKey)

	serverSecretKey, err := ioutil.ReadFile(privateKeysDir + "/server.key_secret")
	if err != nil {
		glog.Errorf("Error reading certs : %v", err)
		return "", "", "", ""
	}
	serverSecret := string(serverSecretKey)

	return clientPublic, clientSecret, serverPublic, serverSecret
}

// Init method is used to initialize the socket
func Init(address string, publicKeysDir string, privateKeysDir string) (*zmq.Socket, error) {

	if publicKeysDir == "" && privateKeysDir == "" {

		//  Create socket for subscribe
		subscriber, err := zmq.NewSocket(zmq.SUB)
		if err != nil {
			glog.Errorf("Error creating socket : %v ", err)
			return nil, err
		}

		// Connect to the socket
		errSoc := subscriber.Connect(address)
		if errSoc != nil {
			glog.Errorf("Error connecting to socket : %v", errSoc)
			return nil, errSoc
		}

		return subscriber, nil
	}

	//  Create socket for subscribe
	subscriber, err := zmq.NewSocket(zmq.SUB)
	if err != nil {
		glog.Errorf("Error creating socket : %v ", err)
		return nil, err
	}

	// Load required certs
	clientPublic, clientSecret, serverPublic, _ := loadCertificates(publicKeysDir, privateKeysDir)

	subscriber.ClientAuthCurve(serverPublic, clientPublic, clientSecret)

	// Connect to the socket
	errSoc := subscriber.Connect(address)
	if errSoc != nil {
		glog.Errorf("Error connecting to socket : %v", errSoc)
		return nil, errSoc
	}

	return subscriber, nil
}

// Recv method is used to receive the meta-data and image
func Recv(subscriber *zmq.Socket, topic string) (string, string, error) {

	// Set topic to subscribe
	err := subscriber.SetSubscribe(topic)
	if err != nil {
		glog.Errorf("Error while setting topic : %v", err)
		return "", "", err
	}

	//  Receive the topic
	address, err := subscriber.Recv(0)
	if err != nil {
		glog.Errorf("Error while receiving topic %s : %v", address, err)
		return "", "", err
	}

	//  Receive the meta-data and frame
	contents, err := subscriber.RecvMessage(0)
	if err != nil {
		glog.Errorf("Error while receiving JSON & buffer : %v", err)
		return "", "", err
	}

	return contents[0], contents[1], nil
}

// Close method is used to close the socket
func Close(subscriber *zmq.Socket) {
	subscriber.Close()
}

File: libs/ZmqLib/py/sub/subscriber.py
import zmq
import json
import os
import sys
import zmq.auth
from zmq.auth.thread import ThreadAuthenticator


class ZMQ_SUBSCRIBER(object):
    """
    The ZMQ_SUBSCRIBER object to subscriber over a
    desired topic
    """
    def __init__(self, address, public_keys_dir, secret_keys_dir):
        """
        Constructor to initialize the socket
        and create context
        """
        if public_keys_dir == "" and secret_keys_dir == "":
            self.dev_mode = True
        else:
            self.dev_mode = False

        if self.dev_mode:
            self.context = zmq.Context()
            self.subscriber = self.context.socket(zmq.SUB)
            self.subscriber.connect(address)
        else:
            # Check if certs exist
            if not (os.path.exists(public_keys_dir) and
                    os.path.exists(secret_keys_dir)):
                print("Certificates are missing - run generate_certificates.py\
                    script first")
                sys.exit(1)

            # Prepare our context and suscriber
            self.context = zmq.Context()

            # Start an authenticator for this context.
            self.auth = ThreadAuthenticator(self.context)
            self.auth.start()

            # Tell authenticator to use the certificate in a directory
            self.auth.configure_curve(domain='*', location=public_keys_dir)
            self.subscriber = self.context.socket(zmq.SUB)

            # We need two certificates, one for the client and one for
            # the server. The client must know the server's public key
            # to make a CURVE connection.
            client_secret_file = os.path.join(secret_keys_dir,
                                              "zmq_client.key_secret")
            client_public, client_secret =\
                zmq.auth.load_certificate(client_secret_file)

            # Setting client server and public keys
            self.subscriber.curve_secretkey = client_secret
            self.subscriber.curve_publickey = client_public

            server_public_file = os.path.join(public_keys_dir, "server.key")
            server_public, _ = zmq.auth.load_certificate(server_public_file)

            # The client must know the server's public key to make a
            # CURVE connection.
            self.subscriber.curve_serverkey = server_public
            self.subscriber.connect(address)

    def receive(self, topic):
        """
        This method is used to receive the json meta-data
        and numpy frame

        Parameters :
        topic : Topic for subscription

        Returns :
        meta_data       : JSON serialized string of meta-data
        frame_blob      : The published frame from ZMQ bus
        """
        # Select what topic to subscribe
        self.subscriber.setsockopt_string(zmq.SUBSCRIBE, topic)

        # Flag to check last message
        flags = 0

        # Receive the topic
        topic = self.subscriber.recv_string(flags=flags)

        # Receive the JSON meta-data
        metaData = self.subscriber.recv_json(flags=flags)

        # Receive the numpy frame
        frame_blob = self.subscriber.recv(flags=flags)

        # Printing the subscribed meta-data and topic
        print("Received message {0} on topic {1}".format(metaData, topic))

        return metaData, frame_blob

    def receive_multipart(self, topic):
        """
        This method is used to receive data in multipart

        Parameters :
        topic : Topic for subscription

        Returns :
        meta_data       : JSON serialized string of meta-data
        frame_blob      : The published frame from ZMQ bus
        """
        # Select what topic to subscribe
        self.subscriber.setsockopt_string(zmq.SUBSCRIBE, topic)

        # Flag to check last message
        flags = 0

        # Receive in multipart
        parts = self.subscriber.recv_multipart(flags=flags)

        # Printing the subscribed meta-data and topic
        print("Received message {0} on topic {1}".format(parts[1], parts[0]))

        return parts[1], parts[2]

    def close_socket(self):
        # Close the socket and destroy the context
        self.subscriber.close()
        self.context.term()
        if not self.dev_mode:
            self.auth.stop()
